magenta-lib/src/test/scala/magenta/deployment_type/CloudFormationTest.scala

```diff
@@ -119,11 +119,118 @@ class CloudFormationTest extends FlatSpec with Matchers with Inside with EitherV
+    val existingParameter =
+      List(ExistingParameter("param1", "monkey", None), ExistingParameter("param2", "monkey", None))
+    val templateParameters =
+      List(TemplateParameter("param1", default = true), TemplateParameter("param2", default = true))
+    val combined = combineParameters(deployParameters, existingParameter, templateParameters, Map.empty)
+
+    combined.right.value should be(Map(
+      "param1" -> UseExistingValue,
+      "param2" -> UseExistingValue,
+    ))
+  }
+
+  it should "allow default values when updating a stack with new parameters" in {
+    val deployParameters = Map("Stack" -> "cfn", "Stage" -> "PROD", "BuildId" -> "543")
+    val existingParameter =
+      List(ExistingParameter("param1", "monkey", None), ExistingParameter("param2", "cheese", None))
+    val templateParameters =
+      List(TemplateParameter("param1", default = true), TemplateParameter("param2", default = false), TemplateParameter("param3", default = true))
+    val combined = combineParameters(deployParameters, existingParameter, templateParameters, Map.empty)
+
+    combined.right.value should be(Map(
+      "param1" -> UseExistingValue,
+      "param2" -> UseExistingValue,
+    ))
+  }
+
+  it should "support new parameters with no default when the user directly supplies values" in {
+    val deployParameters = Map("Stack" -> "cfn", "Stage" -> "PROD", "BuildId" -> "543")
+    val existingParameter =
+      List(ExistingParameter("param1", "monkey", None), ExistingParameter("param2", "cheese", None))
+    val templateParameters =
+      List(TemplateParameter("param1", default = true), TemplateParameter("param2", default = false), TemplateParameter("param3", default = false))
+    val combined = combineParameters(deployParameters, existingParameter, templateParameters, Map("param3" -> "user-provided"))
+
+    combined.right.value should be(Map(
+      "param1" -> UseExistingValue,
+      "param2" -> UseExistingValue,
+      "param3" -> SpecifiedValue("user-provided")
+    ))
+  }
+
+  it should "always specify parameters when the user directly supplies values" in {
+    val deployParameters = Map("Stack" -> "cfn", "Stage" -> "PROD", "BuildId" -> "543")
+    val existingParameter =
+      List(ExistingParameter("param1", "monkey", None), ExistingParameter("param2", "cheese", None), ExistingParameter("param3", "something-else", None))
+    val templateParameters =
+      List(TemplateParameter("param1", default = true), TemplateParameter("param2", default = false), TemplateParameter("param3", default = false))
+    val combined = combineParameters(deployParameters, existingParameter, templateParameters, Map("param3" -> "user-provided"))
+
+    combined.right.value should be(Map(
+      "param1" -> UseExistingValue,
+      "param2" -> UseExistingValue,
+      "param3" -> SpecifiedValue("user-provided")
+    ))
+  }
+
+  it should "always specify parameters when the template contains deploy parameters" in {
```

Comment:
I think this partly shows that riff-raff would have resolved our case in a sensible way with this change - if I'm reading this right the deploy parameter "PROD" is chosen for "Stage" over the existing parameter "CODE".

I think our example was slightly different in that there was no existing parameter (we were deploying the stack for the first time), only a default parameter for "Stage" which was different to the deploy parameter (default was "CODE" but we were deploying "PROD"). A sensible behaviour in this case would have been for the deploy parameter to be chosen over the default parameter.

I'm sorry if I've missed a test which spells this out, or something else which obviously says deploy parameters should be used over anything else.

Reply:
That particular behaviour hasn't changed. If you have a parameter called `Stage` then it would always have been populated with the stage name that Riff-Raff was targetting to. 

As you surmise this test is just trying to prove that the existing values are irrellevant.

Reply:
Excellent